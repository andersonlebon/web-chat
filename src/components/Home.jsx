import React, { useState,  useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Link } from 'react-router-dom';
import Conversation from './chats/chatRoom';
import { setUser, setUsers } from '../store/users/user';
import AllUsers from './chats/chatSideBar';






const Home = () => {
    const dispatch = useDispatch()
    const { chats } = useSelector(state => state.chat);
    const user = JSON.parse(sessionStorage.getItem('user')) || null;
   
    useEffect(() => {
        const users = JSON.parse(localStorage.getItem('users')) || [];
        dispatch(setUsers(users))
    }, [dispatch]);
    // const usersAll = users.filter(item => item.id !== user.id);

    return (
        <main className="homepage">
            <AllUsers/>
            <h1>Welcome to the Web chat</h1>
            <ul className="">
            { chats.map(chat => <li key={chat.id}><Link to={`/conversation/${user.id !== chat.user1 ? chat.user1 : chat.user2 }`} key={chat.id}>{user.id !== chat.user1 ? chat.username1 : chat.username2 }</Link></li>)}
            </ul>            
            
        </main>
    )
}
 
export default Home;