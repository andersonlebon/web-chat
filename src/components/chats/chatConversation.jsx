import React from 'react';
import ChatMessage from '../common/chatMessage';


const Conversation = () => {
    return ( 
        <section className="conversation d-flex flex-column">
            <ChatMessage position="left"/>
            <ChatMessage position="right"/>
            <ChatMessage position="left"/>
            <ChatMessage position="right"/>
            <ChatMessage position="left"/>
            <ChatMessage position="right"/>
            <ChatMessage position="left"/>
            <ChatMessage position="right"/>
            <ChatMessage position="left"/>
            <ChatMessage position="left"/>

            <ChatMessage position="left"/>

            <ChatMessage position="left"/>

            <ChatMessage position="right"/>
        </section>
     );
}
 
export default Conversation;