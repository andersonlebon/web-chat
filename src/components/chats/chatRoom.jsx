import React, { useEffect, useState} from 'react';
import { useSelector, useDispatch } from 'react-redux';
import ChatMessage from '../common/chatMessage';
import { setMessages, saveMessage} from '../../store/chat/chat';
import Input from '../common/input';
import { v4 as uuid } from 'uuid';
import { useParams } from 'react-router';
import { MdSend } from 'react-icons/md';



const Conversation = () => {
    const [message, setMessage] = useState('');
    const { chat, users } = useSelector(state => state);
    const user = JSON.parse(sessionStorage.getItem('user')) || {};
    const { id } = useParams();
    const dispatch = useDispatch();
    useEffect(() => {
        const messages = JSON.parse(localStorage.getItem('messages')) || [];
        dispatch({type: setMessages.type, payload: messages});
    }, [dispatch]);
    const{ messages }  = chat.chats.find(item => item.user1 === user.id && item.user2 === id) || chat.chats.find(item => item.user2 === user.id && item.user1 === id) || {};
    console.log(messages);


     const handleChange = (e) => {
        setMessage(e.target.value);
    }
    const handleSubmit = (e) => {
        e.preventDefault();
        let today = new Date();
        let date = today.getFullYear() + '-' + (today.getMonth() + 1) + '-' + today.getDate();
        const time = today.getHours() + ":" + today.getMinutes() + ":" + today.getSeconds();
        console.log(id);

        const amessage = {
            id: uuid(),
            authorId: user.id,
            message,
            time: time,
            date: date, 
        };

        console.log(user.id);

        const messages = JSON.parse(localStorage.getItem('messages')) || [];
            const newMessage = messages.find(item => item.user1 === user.id && item.user2 === id) || messages.find(item => item.user2 === user.id && item.user1 === id)
            if (newMessage) {
                newMessage.messages.push(amessage);
                console.log(newMessage);
    }
        else {
            messages.push({
                id: uuid(),
                user1: user.id,
                user2: id,
                username1: user.name,
                username2: users.users.find(item => item.id === id).name,
                messages: [amessage],
                sender :amessage.sender || "",
                receiver :amessage.receiver || null,
                createdAt : { time: amessage.time, 
                    date: amessage.date,
                }
            })
        }
        localStorage.setItem('messages', JSON.stringify(messages));
        dispatch({type: setMessages.type, payload : messages}) 
        setMessage('');
    }

    
    return ( 
        <section className="conversation d-flex flex-column ">
            { messages? messages.map(message => <ChatMessage key={message.id} message={message} position={user.id === message.authorId ? 'right' : 'left'}/>) : 'no conversations'}
            <form className="d-flex bg-primary justify-content-between" onSubmit={handleSubmit}>
                <Input type="text" clas="form-control" onchange={handleChange} value={message} name="message"/>
                <button type="submit" className="text-primary send-btn"><MdSend/></button> 
            </form>
            </section>
     );
}
 
export default Conversation;