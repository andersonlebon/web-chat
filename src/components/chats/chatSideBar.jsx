import React, { useEffect} from 'react';
import { useSelector } from 'react-redux'
import { Link } from 'react-router-dom'
import { FaUserCircle } from 'react-icons/fa'



const AllUsers = () => {
    const { users } = useSelector(state => state.users)
    const user = JSON.parse(sessionStorage.getItem('user')) || null;
    const usersAll = users.filter(item => item.id !== user.id) || {};

    
    return ( 
        <section className="side-chat bg-light">
            <ul className="m-0 p-0">
            {users.map(auser => (
                <li  key={auser.id} className="d-flex bg-white">
                    <Link className="d-flex align-items-center justify-content-between" to={`/conversation/${auser.id}`} key={auser.id}>
                        <div className="name-icon d-flex align-items-center justify-content-between">
                            <div className="icon-img">
                                <FaUserCircle/>
                            </div>
                        <h2 className="name pl-3 text-secondary"> {auser.name}</h2>
                        </div>

                        <div className="date-time">
                            <p className="date">{auser.time || "2020-20-10"}</p>
                            <p className="time">{auser.date || "10:10:10"}</p>
                        </div>
                    </Link>
                </li>))}
            </ul>

        </section>
     );
}
 
export default AllUsers;