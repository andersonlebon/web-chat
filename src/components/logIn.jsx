import React, { useState, useEffect} from 'react';
import { useDispatch } from 'react-redux';
import { useNavigate } from 'react-router';
import { Link } from 'react-router-dom';
import { setUser } from '../store/users/user';
import Input from './common/input';


const LogIn = () => {
    const [nameValue, setnameValue] = useState('');
    const dispatch = useDispatch();
    const navigate = useNavigate()

    const handleChange = (e) => {
        setnameValue(e.target.value);
    }
    const handleSubmit = (e) => {
        e.preventDefault();
         const users = JSON.parse(localStorage.getItem('users')) || [];
        const user = users.find(item => item.name === nameValue);

        if (user) {
            sessionStorage.setItem('user', JSON.stringify(user));
            dispatch({ type: setUser.type, payload: user });
            navigate(`/`)

        }
        else {
        dispatch({type: setUser.type, payload: {name: nameValue}});
        }

        setnameValue('');
    }

    return ( 
        <section className="logIn d-flex flex-column justify-content-center" onSubmit={handleSubmit}>
            <h1 className="text-center weight-bold">Log In</h1>
            <form className="flex-column d-flex p-3">
                <div className="group-control my-5">
                <label htmlFor="liginInput" className="title">Name</label>
                <Input type="text" id="liginInput" onchange={handleChange} value={nameValue} clas="form-control" name="username"/>
                
                </div>
                <button className="btn btn-primary" type="submit">Sign In</button>
                <span className="text-center text-secondary">or</span>
                <Link to="/signup" className="btn btn-primary" type="submit">Sign Up</Link>

            </form>
        </section>
                
     );
}
 
export default LogIn;