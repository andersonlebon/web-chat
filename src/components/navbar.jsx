import React, { useState, useEffect } from 'react';
import { FaUserCircle } from 'react-icons/fa';
import { MdOutlineLogout } from 'react-icons/md';
import { Link } from 'react-router-dom';


const Navbar = ({currentUser}) => {
    return ( 
    <header className="navbar-chat d-flex  justify-content-between align-items-center">
        <div className="navbar-container w-100 d-flex align-items-center justify-content-between">
            <div className="navbar-menu">
                <ul className="d-flex  justify-content-center align-items-center">
                    <li className="navbar-menu-item">
                        <Link to="/">Home</Link>
                    </li>
                    <li className="navbar-menu-item">
                        <Link to="/newconversation">new conversation</Link>
                    </li>
                </ul>
            </div>
            <div className="navbar-logo d-flex flex-column align-items-center justify-content-around">
                <div className="icon text-primary"><FaUserCircle /></div> <div>{currentUser.name || ''}</div>
            </div>
            <button className="btn log-out btn-primary"><MdOutlineLogout/></button>
        </div>
    </header>
    );

}
 
export default Navbar;