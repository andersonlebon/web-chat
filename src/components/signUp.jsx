import React, { useState } from 'react';
import Input from './common/input';
import { useDispatch, useSelector } from 'react-redux';
import { saveUser } from '../store/users/user';
import { v4 as uuid } from 'uuid';
import { Link, useNavigate } from 'react-router-dom';

const SignUp = () => {
    const [nameValue, setnameValue] = useState('');
    const { error } = useSelector(state => state.users)
    const dispatch = useDispatch()
    const navigate = useNavigate()
    const handelChange = (e) => {
        setnameValue(e.target.value);
    }
    const handelSubmit = (e) => {
        e.preventDefault();
        const users = JSON.parse(localStorage.getItem('users')) || [];
        const newUser = users.find(item => item.name === nameValue);
        const today = new Date();
        const date = today.getFullYear() + '-' + (today.getMonth() + 1) + '-' + today.getDate();
        const time = today.getHours() + ":" + today.getMinutes() + ":" + today.getSeconds();
        const user = {
                name: nameValue,
                id: uuid(),
                time: time,
                date: date, 
        }
        if (!newUser) {
            users.push(user);
            localStorage.setItem('users', JSON.stringify(users));
            sessionStorage.setItem('user', JSON.stringify(user));
            dispatch({type: saveUser.type, payload: user});
            navigate(`/`)
            window.location.reload();
            

        }

        dispatch({type: saveUser.type, payload: user});

        
    }
    return (
    <section className="logIn d-flex flex-column justify-content-center" onSubmit={handelSubmit}>
            <h1 className="text-center weight-bold">Sign Up</h1>
            <form className="flex-column d-flex p-3">
                <div className="group-control my-4">
                <label htmlFor="liginInput" className="title">Name</label>
                <Input type="text" id="liginInput" onchange={handelChange} value={nameValue} clas="form-control" name="username"/>
                {error && <p className="alert-danger px-2 py-0 my-2 alert">{error}</p>}

                </div>
                <button className="btn btn-primary" type="submit">Sign Up</button>
                <span className="text-center text-secondary">or</span>
                <Link to="/login" className="btn btn-primary" type="submit">Sign In</Link>

            </form>
        </section>
    
    );
}
 
export default SignUp;