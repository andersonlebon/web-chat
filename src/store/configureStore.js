import { combineReducers } from "redux";
import { configureStore } from "@reduxjs/toolkit";
import chatReducer from './chat/chat';


const rootReducer = combineReducers({
    chatReducer,
});

const store = configureStore({ reducer: rootReducer });

export default store;