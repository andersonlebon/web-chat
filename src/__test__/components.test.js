// test reducres
import React from 'react';
import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { BrowserRouter, Route, Routes } from 'react-router-dom';
import { Provider, useSelector } from 'react-redux';
import store from '../store/configureStore';
import { Conversation } from '../components/chats/chatRoom';
import { AllUsers } from '../components/chats/chatSideBar';
import { Home } from '../components/Home';

const MockApp = () => {
    return (<BrowserRouter>
    <Provider store={store}>
        <Routes>
            <Route path="/" element={Home}/>
            <Route path="/conversation/:id" element={Conversation}/>
            <Route path="/newconversation" element={AllUsers}/>
        </Routes>
    </Provider>
    </BrowserRouter>)
};

describe('test Conversation component', () => {
    beforeEach(() => {
        render(<MockApp />);

    });
    it('should render correctly', () => {
        const tree = renderer.create(<MockApp />).toJSON();
        expect(tree).toMatchSnapshot();
    });
    
}
);


