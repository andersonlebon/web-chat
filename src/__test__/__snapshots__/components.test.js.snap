// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Conversation component should render correctly 1`] = `null`;
